### ISSUE TITLE
XCode Build Warnings

### ISSUE THREAD

**Issue by jwmelto** [BZFlag-Dev/bzflag#176]
XCode build produces a number of warnings. Some include:

ld: warning: direct access in function 'AccessControlList::sendIdBans(unsigned char, char const*)' from file '/Users/jwmelto/Library/Developer/Xcode/DerivedData/BZFlag-baadnscasurjbjhjxacwopmhceik/Build/Intermediates.noindex/BZFlag.build/Debug/bzfs.build/Objects-normal/x86_64/AccessControlList.o' to global weak symbol 'std::__1::char_traits<char>::eq(char, char)' from file '/Users/jwmelto/Library/Developer/Xcode/DerivedData/BZFlag-baadnscasurjbjhjxacwopmhceik/Build/Products/Debug/libobstacle.a(ObstacleMgr.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.

[This post](
https://stackoverflow.com/questions/8685045/xcode-with-boost-linkerid-warning-about-visibility-settings) suggests that a build setting can silence those warnings. 

Other warnings are more benign and just indicate a lack of discipline in code quality. The project should build without warnings.

**Comment by macsforme** [comment]
These two compiler flags clear out all our compiler warnings in Xcode (other than for some deprecated functions we use, mostly for GLU):

    -fvisibility=hidden
    -Wno-undefined-var-template

The former does cause plug-ins to fail to load, so the setting might not apply to the bzfs project, or we might need to mess with the visibility macros so it works like on Windows ([this](https://stackoverflow.com/questions/37934388/symbol-visibility-not-working-as-expected) seems like a good place to start):

```
Player macsforme [0] -> All: /loadplugin airspawn.dylib
Plugin: airspawn.dylib not found, error dlopen(airspawn.dylib, 9): Symbol not found: __ZTI9bz_Plugin
  Referenced from: airspawn.dylib
  Expected in: flat namespace
 in airspawn.dylib
```

The latter is a different warning than the one described in the first post (although this one is also very prevalent):

```
In file included from src/bzflag/EffectsMenu.cxx:18:
In file included from include/FontManager.h:20:
include/Singleton.h:97:14: warning: instantiation of variable 'Singleton<EffectsRenderer>::_instance' required here, but no definition is available [-Wundefined-var-template]
        if ( _instance == 0 )
             ^
src/bzflag/EffectsMenu.cxx:156:14: note: in instantiation of member function 'Singleton<EffectsRenderer>::instance' requested here
    optbuf = EFFECTS.getSpawnEffectTypes();
             ^
In file included from src/bzflag/EffectsMenu.cxx:26:
src/bzflag/effectsRenderer.h:39:35: note: expanded from macro 'EFFECTS'
#define EFFECTS (EffectsRenderer::instance())
                                  ^
In file included from src/bzflag/EffectsMenu.cxx:18:
In file included from include/FontManager.h:20:
include/Singleton.h:62:15: note: forward declaration of template entity is here
    static T* _instance;
              ^
include/Singleton.h:97:14: note: add an explicit instantiation declaration to suppress this warning if 'Singleton<EffectsRenderer>::_instance' is explicitly instantiated in another translation unit
        if ( _instance == 0 )
             ^
```

Some quick info found online [here](https://github.com/dealii/dealii/issues/3705) and [here](https://github.com/InsightSoftwareConsortium/ITKIsotropicWavelets/issues/120).

**Comment by jwmelto** [comment]
I'm not sure how you came up with `-fvisibility=hidden` but that would seem to a very bad idea, at this point.

There might be some value at some point in requiring explicit export of symbols, but that's far afield from silencing warnings.

**Comment by jwmelto** [comment]
#203

**Comment by jwmelto** [comment]
>>ld: warning: direct access in function 'AccessControlList::sendIdBans(unsigned char, char const*)' from file '/Users/jwmelto/Library/Developer/Xcode/DerivedData/BZFlag-afjcbekkkbrjoebykdssrwuhbpua/Build/Intermediates.noindex/BZFlag.build/Debug/bzfs.build/Objects-normal/x86_64/AccessControlList.o' to **global weak symbol 'std::__1::char_traits<char>::eq(char, char)'** from file '/Users/jwmelto/Library/Developer/Xcode/DerivedData/BZFlag-afjcbekkkbrjoebykdssrwuhbpua/Build/Products/Debug/libobstacle.a(ObstacleMgr.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.

I'm not clear on weak symbols, but `std::char_traits<char>::eq(char,char)` looks like maybe a template has crossed a DLL boundary?

Then again [this](https://stackoverflow.com/a/28827430/9220132) suggests that it may just be a difference between GCC and Clang?

**Comment by macsforme** [comment]
The `-fvisibility=hidden` was I believe the suggested remedy in the post you linked to in the original PR. It was suggested that the issue was conflicting visibility settings. How this is happening, I don’t know... from what I can tell, it’s all internal to our code.

**Comment by jwmelto** [comment]
you said `-fvisibility=hidden` caused plugins to fail to load; @blast007 suggested that proper application of BZF_API however it is defined differently for Windows than unices. However, if we explicitly set visibility to hidden **and** change the definition of BZF_API for non-Windows to include an `__attribute__((visibility("default")))` (or possibly a C++11 `[[gnu::visibility("default")]]` to the definition.

For pessimism's sake, we could only do this for **clang** and leave GCC to merrily do whatever it's doing that happens to work.

**Comment by macsforme** [comment]
When we have `GCC_SYMBOLS_PRIVATE_EXTERN = YES`, `GCC_INLINES_ARE_PRIVATE_EXTERN = YES`, and `ENABLE_TESTABILITY = NO` (which avoids an issue with the visibility settings being overridden in debug mode), all the "direct access" warnings go away, but we get:

    $ ./bzfs -loadplugin airspawn.dylib
    Plugin: airspawn.dylib not found, error dlopen(airspawn.dylib, 9): Symbol not found: __ZTI9bz_Plugin

When we `#define BZF_API __attribute__((visibility("default")))`, now we get:

    $ ./bzfs -loadplugin airspawn.dylib
    Plugin: airspawn.dylib found but does not contain bz_GetPlugin method, error dlsym(0x7f906d6113f0, bz_GetPlugin): symbol not found

When we also `#define BZF_PLUGIN_CALL extern "C" __attribute__((visibility("default")))`, we get:

    $ ./bzfs -loadplugin airspawn.dylib
    Plugin: Airspawn loaded from airspawn.dylib

I'm way outside of my comfort zone here, but trying to imitate the way it works on Windows, I got a patch that seems to make it all work:

```
--- a/Xcode/BZFlag.xcodeproj/project.pbxproj
+++ b/Xcode/BZFlag.xcodeproj/project.pbxproj
@@ -6402,7 +6402,6 @@
                                        "INSIDE_BZ=1",
                                        "DEBUG=1",
                                );
-                               GCC_SYMBOLS_PRIVATE_EXTERN = NO;
                                PRODUCT_NAME = "$(TARGET_NAME)";
                        };
                        name = Debug;
@@ -6414,7 +6413,6 @@
                                        "$(inherited)",
                                        "INSIDE_BZ=1",
                                );
-                               GCC_SYMBOLS_PRIVATE_EXTERN = NO;
                                PRODUCT_NAME = "$(TARGET_NAME)";
                        };
                        name = Release;
@@ -7292,11 +7290,12 @@
                                CLANG_WARN_EMPTY_BODY = YES;
                                CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
                                COPY_PHASE_STRIP = NO;
-                               ENABLE_TESTABILITY = YES;
+                               ENABLE_TESTABILITY = NO;
                                FRAMEWORK_SEARCH_PATHS = /Library/Frameworks;
                                GCC_C_LANGUAGE_STANDARD = gnu99;
                                GCC_DYNAMIC_NO_PIC = NO;
                                GCC_ENABLE_OBJC_EXCEPTIONS = YES;
+                               GCC_INLINES_ARE_PRIVATE_EXTERN = YES;
                                GCC_OPTIMIZATION_LEVEL = 0;
                                GCC_PREPROCESSOR_DEFINITIONS = (
                                        "$(inherited)",
@@ -7309,7 +7308,7 @@
                                        HAVE_SDL2_SDL_H,
                                        "DEBUG=1",
                                );
-                               GCC_SYMBOLS_PRIVATE_EXTERN = NO;
+                               GCC_SYMBOLS_PRIVATE_EXTERN = YES;
                                GCC_WARN_64_TO_32_BIT_CONVERSION = NO;
                                GCC_WARN_ABOUT_RETURN_TYPE = YES;
                                GCC_WARN_UNINITIALIZED_AUTOS = YES;
@@ -7337,9 +7336,11 @@
                                CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
                                COPY_PHASE_STRIP = YES;
                                DEBUG_INFORMATION_FORMAT = "dwarf-with-dsym";
+                               ENABLE_TESTABILITY = NO;
                                FRAMEWORK_SEARCH_PATHS = /Library/Frameworks;
                                GCC_C_LANGUAGE_STANDARD = gnu99;
                                GCC_ENABLE_OBJC_EXCEPTIONS = YES;
+                               GCC_INLINES_ARE_PRIVATE_EXTERN = YES;
                                GCC_PREPROCESSOR_DEFINITIONS = (
                                        "$(inherited)",
                                        HAVE_CONFIG_H,
@@ -7350,6 +7351,7 @@
                                        HAVE_SDL2,
                                        HAVE_SDL2_SDL_H,
                                );
+                               GCC_SYMBOLS_PRIVATE_EXTERN = YES;
                                GCC_WARN_64_TO_32_BIT_CONVERSION = NO;
                                GCC_WARN_ABOUT_RETURN_TYPE = YES;
                                GCC_WARN_UNINITIALIZED_AUTOS = YES;
diff --git a/include/bzfsAPI.h b/include/bzfsAPI.h
index 1acf156cbd..2844e92404 100644
--- a/include/bzfsAPI.h
+++ b/include/bzfsAPI.h
@@ -39,9 +39,14 @@
 #define strcasecmp stricmp
 #endif
 #else
+#ifdef __clang__
+#define BZF_API __attribute__((visibility("default")))
+#define BZF_PLUGIN_CALL extern "C" __attribute__((visibility("default")))
+#else
 #define BZF_API
 #define BZF_PLUGIN_CALL extern "C"
 #endif
+#endif
 
 /* Provide a means to deprecate API functions to discourage their use
  * in the future
```
Let me know how that looks.

**Comment by macsforme** [comment]
This issue was resolved with the combination of pull requests #203, #206, and #207.
